src/Database/Esqueleto/Record.hs

```diff
@@ -0,0 +1,423 @@
+--         myUser         <- 'takeColumns' @('SqlExpr' ('Entity' User))
+--         myAddress      <- 'takeColumns' @('SqlExpr' ('Maybe' ('Entity' Address)))
+--         pure MyRecord { myName = myName
+--                       , myAge = myAge
+--                       , myUser = myUser
+--                       , myAddress = myAddress
+--                       }
+-- @
+--
+-- Then, we could write a selection function to use the record in queries:
+--
+-- @
+-- getMyRecord :: SqlPersistT IO [MyRecord]
+-- getMyRecord = select myRecordQuery
+--
+-- myRecordQuery :: SqlQuery SqlMyRecord
+-- myRecordQuery = do
+--   user :& address <- from $
+--     table @User
+--       `leftJoin`
+--       table @Address
+--       `on` (do \(user :& address) -> user ^. #address ==. address ?. #id)
+--   pure
+--     SqlMyRecord
+--       { myName = castString $ user ^. #firstName
+--       , myAge = val 10
+--       , myUser = user
+--       , myAddress = address
+--       }
+-- @
+--
+-- @since 3.6.0.0
+deriveEsqueletoRecord :: Name -> Q [Dec]
+deriveEsqueletoRecord originalName = do
+  info <- getRecordInfo originalName
+  mconcat
+    [ makeSqlRecord info
+    , makeSqlSelectInstance info
+    ]
+
+-- | Information about a record we need to generate the declarations.
+-- We compute this once and then pass it around to save on complexity /
+-- repeated work.
+data RecordInfo = RecordInfo
+  { -- | The original record's name.
+    name :: Name
+  , -- | The generated @Sql@-prefixed record's name.
+    sqlName :: Name
+  , -- | The original record's constraints. If this isn't empty it'll probably
+    -- cause problems, but it's easy to pass around so might as well.
+    constraints :: Cxt
+  , -- | The original record's type-variable-binders.
+    typeVarBinders :: [TyVarBndr ()]
+  , -- | The original record's kind, I think.
+    kind :: Maybe Kind
+  , -- | The original record's constructor name.
+    constructorName :: Name
+  , -- | The original record's field names and types, derived from the
+    -- constructors.
+    fields :: [(Name, Type)]
+  , -- | The generated @Sql@-prefixed record's field names and types, computed
+    -- with 'sqlFieldType'.
+    sqlFields :: [(Name, Type)]
+  }
+
+-- | Get a `RecordInfo` instance for the given record name.
+getRecordInfo :: Name -> Q RecordInfo
+getRecordInfo name = do
+  TyConI dec <- reify name
+  let (constraints, typeVarBinders, kind, constructors) =
+        case dec of
+          DataD constraints' _name typeVarBinders' kind' constructors' _derivingClauses ->
+            (constraints', typeVarBinders', kind', constructors')
+          NewtypeD constraints' _name typeVarBinders' kind' constructor' _derivingClauses ->
+            (constraints', typeVarBinders', kind', [constructor'])
+          _ -> error $ "Esqueleto records can only be derived for records and newtypes, but " ++ show name ++ " is neither"
+      constructor = fromJust $ listToMaybe constructors
```

Comment:
We may want to fail with a more sensible error here - believe it or not, you can write:

```haskell
data Void
```

which has no constructors.

```suggestion
    constructor <- 
      case constructors of
        (constr : _) -> pure constr
        [] -> fail "Cannot derive esqueleto record for a type with no constructors"
```

🤔